src/hotspot/share/services/memReporter.hpp

```diff
@@ -66,11 +66,22 @@ class MemReporterBase : public StackObj {
   }
 
   // Convert diff amount in bytes to current reporting scale
-  inline long diff_in_current_scale(size_t s1, size_t s2) const {
-    long amount = (long)(s1 - s2);
-    long scale = (long)_scale;
-    amount = (amount > 0) ? (amount + scale / 2) : (amount - scale / 2);
-    return amount / scale;
+  inline int64_t diff_in_current_scale(size_t s1, size_t s2) const {
+    // _scale should not be 0, otherwise division by zero at return.
+    assert(_scale != 0, "wrong scale");
+
+    bool is_negative = false;
+    if (s1 < s2) {
+      is_negative = true;
+      swap(s1, s2);
+    }
+
+    size_t amount = s1 - s2;
+    assert(amount <= SIZE_MAX - _scale / 2, "size_t overflow");
+    amount = (amount + _scale / 2) / _scale;
+    // We assume the valid range for deltas [INT64_MIN, INT64_MAX] to simplify the code.
+    assert((sizeof(size_t) < sizeof(int64_t)) || (sizeof(size_t) == sizeof(int64_t) && amount <= INT64_MAX), "cannot fit scaled diff into size_t");
```

Comment:
Maybe this would better:
"scaled diff is out of supported range [-INT64_MAX, INT64_MAX]"